### PR TITLE
TaurusValuesTable optional slices (number of items to show / row)

### PR DESCRIPTION
**Pull request by jairomoldes**
This feature allows to optionally display 1D data as a table and optionally display different number of items in each row

Sample usage:

<pre>
#!/usr/bin/env python

def main():
    import sys
    from taurus.qt.qtgui.application import TaurusApplication

    app = TaurusApplication()
    form = Qt.QWidget()
    layout = Qt.QVBoxLayout()
    form.setLayout(layout)
    from taurus.qt.qtgui.table import TaurusValuesTable
    w = TaurusValuesTable()
    w.setModel('sys/tg_test/1/boolean_spectrum')
    w.setWriteMode(False)
    _slice = [10 for i in range(10)]
    _slice.extend([5 for i in range(8)])
    _slice.extend([6 for i in range(4)])
    _slice.extend([1,1])
    _slice.extend([9 for i in range(10)])
    w.setSlices(_slice)
    layout.addWidget(w)
    form.show()
    sys.exit(app.exec_())

if __name__ == "__main__":
    main()
</pre>

### CODE REVIEW

**Comment by jairomoldes** [comment]
Sample usage:

<pre>
#!/usr/bin/env python

def main():
    import sys
    from taurus.qt.qtgui.application import TaurusApplication

    app = TaurusApplication()
    form = Qt.QWidget()
    layout = Qt.QVBoxLayout()
    form.setLayout(layout)
    from taurus.qt.qtgui.table import TaurusValuesTable
    w = TaurusValuesTable()
    w.setModel('sys/tg_test/1/boolean_spectrum')
    w.setWriteMode(False)
    _slice = [10 for i in range(10)]
    _slice.extend([5 for i in range(8)])
    _slice.extend([6 for i in range(4)])
    _slice.extend([1,1])
    _slice.extend([9 for i in range(10)])
    w.setSlices(_slice)
    layout.addWidget(w)
    form.show()
    sys.exit(app.exec_())

if __name__ == "__main__":
    main()
</pre>

**Comment by jairomoldes** [comment]
I would like to illustrate what I would like to get with pull requests:
https://github.com/taurus-org/taurus/pull/443
https://github.com/taurus-org/taurus/pull/427

I have an 88 boolean spectrum attribute that represent the 88 BPMs that can be used for our FOFB systems. These 88 BPMs are distributed in the 16 sectors of the machine and furthermore the number of BPMs per sector may be different. I think that this feature would be of general use for other people.

Have a look at the attached images to see the difference between current status and what I propose with my PRs.

boolarray1
boolarray2

**Comment by cmft** [comment]
Testing the example the right click fails when write mode is enabled:

- In empty cells. 
```
  File "/home/cfalcon/tools/taurus.git/lib/taurus/qt/qtgui/table/taurusvaluestable.py", line 872, in contextMenuEvent
    index = self._tableView.selectedIndexes()[0]
IndexError: list index out of range
```

- In cells with value (After triggering the first error).
```
Traceback (most recent call last):
  File "/home/cfalcon/tools/taurus.git/lib/taurus/qt/qtgui/table/taurusvaluestable.py", line 874, in contextMenuEvent
    val = self._tableView.model().getReadValue(index)
  File "/home/cfalcon/tools/taurus.git/lib/taurus/qt/qtgui/table/taurusvaluestable.py", line 384, in getReadValue
    return self._rtabledata[index.row(), index.column()]
IndexError: index 3 is out of bounds for axis 1 with size 1

``` 

The error is not triggered with Develop

File: lib/taurus/qt/qtgui/table/taurusvaluestable.py [hunk 1]
```diff
@@ -384,6 +384,105 @@ def getReadValue(self, index):
...
+    Table model that allows representing 1D data in 2D format with optionally
+    different number of items per row
+    '''
+
+    def __init__(self, slices=[], parent=None):
+        self._parent = parent
+        name = self.__class__.__name__
+        TaurusValuesIOTableModel.__init__(self, [1, 1], parent)
+        self.slices = slices
+        self._super = super(TaurusValuesSlicedArrayModel, self)
```
Comment (cpascual) [inline lib/taurus/qt/qtgui/table/taurusvaluestable.py]: I think that it is better not to keep this hard reference to super. Better use super explicitly, or simply use `TaurusValuesIOTableModel`

File: lib/taurus/qt/qtgui/table/taurusvaluestable.py [hunk 2]
```diff
@@ -384,6 +384,105 @@ def getReadValue(self, index):
...
 
 
+class TaurusValuesSlicedArrayModel(TaurusValuesIOTableModel):
+    '''
+    Table model that allows representing 1D data in 2D format with optionally
+    different number of items per row
+    '''
+
+    def __init__(self, slices=[], parent=None):
+        self._parent = parent
```
Comment (cpascual) [inline lib/taurus/qt/qtgui/table/taurusvaluestable.py]: I am worried about possible circular references. I would try to avoid keeping a hard ref to parent

**Comment by cpascual** [comment]
Also, after testing (I used develop + this PR + PR #473) I found [the same exceptions as @cmft](https://github.com/taurus-org/taurus/pull/443#issuecomment-307816601) when right-clicking on cells in write-mode.
They should be fixed.

**Comment by cpascual** [comment]
@jairomoldes do you think you can address these issues so that we can integrate this for Jul17?

**Comment by jairomoldes** [comment]
I don't think so

**Comment by cpascual** [comment]
Hi, is there any chance of someone working on getting this into shape?
(if not, maybe we should close it as rejected)

**Comment by cpascual** [comment]
No answer, so I am closing. 
Anyone feel free to reopen at any time if you want to work on this feature

**Pull request closed without merging**